feat: LID for query splitting (#16958)
grafana/loki: docs/sources/community/lids/0006-api-expose-split.md

---
title: "0006: Expose Split Logic in API"
description: "0006: Expose Split Logic in API"

> aliases: 
> - ../../lids/0006-api-expose-split/

Reviewer: Since this page hasn't moved or been renamed, you don't need an alias.
```suggestion
```

accessible outside of the code base. This proposal intends to create an API for clients to split queries by exposing the
internal split logic.

## Problem Statement

Loki clients such as the Grafana Loki datasource or the [Trino Loki
connector](https://github.com/trinodb/trino/tree/master/plugin/trino-loki) benefit from splitting LogQL queries into multiple sub-queries either to process
smaller chunks or to distribute work on query results.

Splitting a query requires parsing the LogQL query first but there are no parsers for other languages except Go and
JavaScript.

## Goals

The intended goal is to enable any client to split a query into multiple sub-queries that can be either executed
sequentially or in parallel. The joined result of the sub-queries must be the same as executing the same query.

## Non-Goals

This proposal does not aim to provider pagination for query results.

## Proposals

### Proposal 0: Do nothing

Without an API each client will have to use a LogQL parser.

*Pros*
- The split logic in Loki can be changed at will without breaking client behavior.
- There is no maintanence overhead for an API.

*Cons*

> - Currenlty, the LogQL grammar is specific to Go. It is not easy to port it and the parser to other languages.

Reviewer: ```suggestion
- Currently, the LogQL grammar is specific to Go. It is not easy to port it and the parser to other languages.
```

- Any changes to the splitting logic must be implemented for each client/platform.

### Proposal 1: Expose Splitting in an API

> A new endpoint `GET /loki/api/v1/split_query` is introduced that takes an `splits` parameter and the same parameters as the [/loki/api/v1/query_range](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) endpoint.

Reviewer: ```suggestion
A new endpoint `GET /loki/api/v1/split_query` is introduced that takes a `splits` parameter and the same parameters as the [/loki/api/v1/query_range](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) endpoint.
```

> The `splits` parameter defines the number of desired splits. The API is allowed to return fewer splits than requested.

Reviewer 2: This would imply that the consumer of this endpoint has information about the expected number of splits. This has been one of the main challenges in the current query splitting implementations. For the case of shard splitting, what defines the amount of "splits" are the values of the `__stream_shard__` label. Even with that information, _n_ splits for example, that is not enough for the queries to know the ideal or optimal amount of splits, because the information is not evenly distributed. To get a more accurate number of splits, this should be aggregated with the volume per split, allowing an evenly distributed number of sub-requests to make.

For time splitting the situation is different, where the main driver for the amount of splits is time.

In my experience, I would send a query and a time range to the endpoint and let it decide what's the ideal amount of splits.

Author: > In my experience, I would send a query and a time range to the endpoint and let it decide what's the ideal amount of splits.

That's great feedback. What would dictate the ideal number of splits though? Query volume?

Reviewer 2: Definitely volume, split ~ evenly across sub-requests.

Author: > Although it would be very very interesting to have Loki decide what's the optimal splitting strategy for a given query.

We can iterate. What would be your use case?

Reviewer 3: I agree that, while it definitely increases scope, making this endpoint a bit smarter than just applying the default per-tenant time split, for example, is a really strong argument for doing this work. I would like this endpoint to try and create "optimal" subqueries as well, which we can of course iterate on what that means.

an example use case, in drilldown we have shard splitting, but if you ask for too much you might get an error or a timeout. it would be hard for Loki to know how much data will cause a timeout, but it would be great if we could think of a way for it to make some good educated guesses. in drilldown when we hit the timeout we reduce the number of shards in the the group, so like @matyax said, maybe if we could create groups by volume that would be a good first step?

Reviewer 3: @jeschkies yes, both time splits and adding the `__stream_shard__` label are done client side. The stream shards are grouped, and @matyax wrote logic to retry with a bigger or smaller group based on how quickly the previous response returned, or if it timed out.

> 

Reviewer 2: The parameters I would pass to this endpoint are:
- query (or queries)
- limit, which should represent the total amount of results between the first and last splitted requests
- direction

Reviewer 2: Coming back to the limit, it's a bit more complicated. For metric queries there's no expected limit, as you expect the data for the requested interval. 

For querying logs it's a different story, because you might hit the limit with the first request. For example, let's say you want 5000 logs from the last 3 days, split in 3/6/12/etc. You can easily hit that limit on the first day, so maybe the endpoint should just return a single split.

Author: In my case I often want all logs. So I'll allow `limit=-1` to indicate that all logs should be returned.

Reviewer 3: just to clarify here, is `limit=-1` saying explicitly construct the splits such that each split falls below any loki limits? I think that gets pretty complicated. if instead `limit=-1` is just saying "I don't care about limits", it would probably be simpler to just omit that parameter in that case (rather than using `-1`)

Author: Many users have the issue that they want to query all logs in a time range but do not know how large the log line limit should be. It defaults to 100. There should be a way to tell the split endpoint that I want to get all log lines.

> The response body is JSON encoded

Reviewer: ```suggestion
The response body is JSON encoded:
```

```json
{ 
  "resultType": "matrix" | "streams" | "vector",

>   "links": [

Reviewer 2: I'm curious to know why the sub-requests response has been defined as URLs, which seem more appropriate for pagination than for splitting. I would think of something closer to:

```
type SubRequest {
  start: number;
  end: number;
  limit: number; // is the limit needed?
  query: string;
}
```

E.g.

```
[{
  start: 10,
  end, 200,
  limit: 20,
  query: "{a="b", __stream_shard__~="10|9|8"}"
}]
```

Additionally, the direction should not change between splits, so it shouldn't need to be specified for each split.

Author: Log queries could be sharded in the future.

Reviewer 3: I like including the query in the response as well, primarily for sharding, but I think it gives us more flexibility for how to optimize it.

    {
      "rel": "???",
      "href": "/loki/api/v1/query_range?start=10&end=200&limit=20&direction=forward?query=..."
    },
    {
      "rel": "???",
      "href": "/loki/api/v1/query_range?start=30&end=200&limit=20&direction=forward?query=..."
    }
  ]
}
```

*Pros*
- Clients can split queries independent on the implemation language and platform.

> - Split logic is controlled by Loki and not the client. This means it can be improved e.g. by introducing sharding

Reviewer: ```suggestion
- Split logic is controlled by Loki and not the client. This means it can be improved, for example, by introducing sharding
```

  labels.

*Cons*
- A new API endpoint increases the compatiblity surface area and thus maintanence overhead for Loki maintainers.

### Proposal 2: Support Apache Arrow Flight RPC

Loki could support Apache [Arrow Flight RPC](https://arrow.apache.org/docs/format/Flight.html) which is designed to
exchange large data sets in shards between services.

*Pros*
- Supporting an open standard comes with support for other non-Loki clients.

*Cons*
- Loki would have to support Apache Arrow which make the implementation more complicated.
- Arrow Flight RPC assumes the data is being queried on the first request. Which means all shards are available at the
  same time. However, the intend of this document is that shards can be queried independently.

>   shard

Reviewer: Should this word be here?  

```suggestion
  shard
```

Author: No. That was a mistake. Thanks for reviewing.

**Date:** 03/2025

**Sponsor(s):** @trevorwhitney

**Type:** API

**Status:** Review

**Related issues/PRs:** N/A

**Thread from [mailing list](https://groups.google.com/forum/#!forum/lokiproject):** N/A

---

## Background

Loki has an internal logic to split log and metric queries by time into multiple queries. However, this logic is not
accessible outside of the code base. This proposal intends to create an API for clients to split queries by exposing the
internal split logic.

## Problem Statement

Loki clients such as the Grafana Loki datasource or the [Trino Loki
connector](https://github.com/trinodb/trino/tree/master/plugin/trino-loki) benefit from splitting LogQL queries into multiple sub-queries either to process
smaller chunks or to distribute work on query results.

Splitting a query requires parsing the LogQL query first but there are no parsers for other languages except Go and
JavaScript.

## Goals

The intended goal is to enable any client to split a query into multiple sub-queries that can be either executed
sequentially or in parallel. The joined result of the sub-queries must be the same as executing the same query.

## Non-Goals

> This proposal does not aim to provider pagination for query results.

Reviewer 3: ```suggestion
This proposal does not aim to provide pagination for query results.
```
typo

## Proposals

### Proposal 0: Do nothing

Without an API each client will have to use a LogQL parser.

*Pros*
- The split logic in Loki can be changed at will without breaking client behavior.
- There is no maintanence overhead for an API.

*Cons*
- Currently, the LogQL grammar is specific to Go. It is not easy to port it and the parser to other languages.
- Any changes to the splitting logic must be implemented for each client/platform.

### Proposal 1: Expose Splitting in an API

> A new endpoint `GET /loki/api/v1/split_query` is introduced that takes a `splits` parameter and the same parameters as the [/loki/api/v1/query_range](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) endpoint. The new endoint returns sub-queries split by time.

Reviewer 3: > split-by-time

will this endpoint determine optimal splits? currently the split-by is hardcoded in the loki config per tenant, in which case it seems pretty trivial for a client to be able to arbitrarily create the same splits. is it a goal of this project that this endpoint would be capable of more intelligent splitting rather than just applying the tenant's split-by config?

Author: > will this endpoint determine optimal splits?

It kind of depends what we mean by optimal :slightly_smiling_face: I think the internal sharding layer tries some optimal sharding. Anyhow, we need something to hang on to. Either number of splits, interval or max response size.

The `splits` parameter optionally defines the number of desired splits. The API is allowed to return fewer splits than requested.

The `limit` parameter has extended semantics. Setting it to `-1` for a log stream query indicates to query all logs.

> The response body is JSON encoded:

Reviewer 3: some thoughts on the response. is it important to the user to differentiate between a single split that just happens to be a single split because it was optimal, and a single split because the query could not be split? certain metric queries will not be able to be split in a way they can be re-combined in the client to produce the same result, so should we add a `warnings` field, or some other metadata to communicate this back to the client?

Author: > certain metric queries will not be able to be split in a way they can be re-combined in the client to produce the same resul

That would be sharding. E.g. for certain metric queries. I should specify in the doc that `split` is always by time and `sharding` is by stream labels.

```json
{ 
  "resultType": "matrix" | "streams" | "vector",
  "subqueries": [
    {
      start: <timestamp nanoseconds>,
      end: <timestamp nanoseconds>,
      limit: <number>,
      query: <query string> 
    },
    {
      start: <timestamp nanoseconds>,
      end: <timestamp nanoseconds>,
      limit: <number>,
      query: <query string> 
    }
  ]
}
```

*Pros*
- Clients can split queries independent on the implemation language and platform.

> - Split logic is controlled by Loki and not the client. This means it can be improved, for example, by introducing sharding

Reviewer 3: this is the strongest argument for doing this work IMO. being able to improve split logic to produce optimal splits, as well as centralize what can and can't be split.

  labels.

*Cons*
- A new API endpoint increases the compatiblity surface area and thus maintanence overhead for Loki maintainers.

### Proposal 2: Support Apache Arrow Flight RPC

Loki could support Apache [Arrow Flight RPC](https://arrow.apache.org/docs/format/Flight.html) which is designed to
exchange large data sets in shards between services.

*Pros*
- Supporting an open standard comes with support for other non-Loki clients.

*Cons*
- Loki would have to support Apache Arrow which make the implementation more complicated.
- Arrow Flight RPC assumes the data is being queried on the first request. Which means all shards are available at the

>   same time. However, the intend of this document is that shards can be queried independently.

Reviewer 3: ```suggestion
  same time. However, the intent of this document is that shards can be queried independently.
```

---
title: "0006: Expose Split Logic in API"
description: "0006: Expose Split Logic in API"
---

# 0006: Expose Split Logic in API

**Author:** Karsten Jeschkies (karsten.jeschkies@grafana.com)

**Date:** 03/2025

**Sponsor(s):** @trevorwhitney

**Type:** API

**Status:** Review

**Related issues/PRs:** N/A

**Thread from [mailing list](https://groups.google.com/forum/#!forum/lokiproject):** N/A

---

## Background

Loki has an internal logic to split and shard log and metric queries by time into multiple queries. However, this logic is not
accessible outside of the code base. This proposal intends to create an API for clients to split queries by exposing the
internal split logic.

> A split query is devided by time. The results of a split query can be concatenated in order to form the final

Reviewer 3: ```suggestion
A split query is divided by time. The results of a split query can be concatenated in order to form the final
```

result.

> A sharded query is devided by label values. The results of a sharded cannot be simply concatenated but require some

Reviewer 3: ```suggestion
A sharded query is divided by label values. The results of a sharded cannot always be concatenated but require some
```

extra logic to form the final result. Some queries, such as `topk` cannot be sharded at all.

## Problem Statement

Loki clients such as the Grafana Loki datasource or the [Trino Loki
connector](https://github.com/trinodb/trino/tree/master/plugin/trino-loki) benefit from splitting LogQL queries into multiple sub-queries either to process
smaller chunks or to distribute work on query results.

Splitting a query requires parsing the LogQL query first but there are no parsers for other languages except Go and
JavaScript.

## Goals

The intended goal is to enable any client to split a query into multiple sub-queries that can be either executed
sequentially or in parallel. The joined result of the sub-queries must be the same as executing the same query.

## Non-Goals

This proposal does not aim to provide pagination for query results.

## Proposals

### Proposal 0: Do nothing

Without an API each client will have to use a LogQL parser.

*Pros*
- The split logic in Loki can be changed at will without breaking client behavior.
- There is no maintanence overhead for an API.

*Cons*

> - Currently, the LogQL grammar is specific to Go. It is not easy to port it and the parser to other languages.

Reviewer 3: worth at least calling out there's a JS implementation as well, but I acknowledge it is largely a copy and paste and has a maintenance burden

- Any changes to the splitting logic must be implemented for each client/platform.

### Proposal 1: Expose Splitting in an API

A new endpoint `GET /loki/api/v1/split_query` is introduced that takes a `splits` parameter and the same parameters as the [/loki/api/v1/query_range](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) endpoint. The new endoint returns sub-queries split by time.

The `splits` parameter optionally defines the number of desired splits. The API is allowed to return fewer splits than requested.

> The `limit` parameter has extended semantics. Setting it to `-1` for a log stream query indicates to query all logs.

Author: TIL logcli uses `limit=0`

Reviewer 3: ahh, good to know, lets stick with that then

sequentially or in parallel. The joined result of the sub-queries must be the same as executing the same query.

## Non-Goals

This proposal does not aim to provider pagination for query results.

## Proposals

### Proposal 0: Do nothing

Without an API each client will have to use a LogQL parser.

*Pros*
- The split logic in Loki can be changed at will without breaking client behavior.
- There is no maintanence overhead for an API.

*Cons*
- Currenlty, the LogQL grammar is specific to Go. It is not easy to port it and the parser to other languages.
- Any changes to the splitting logic must be implemented for each client/platform.

### Proposal 1: Expose Splitting in an API

A new endpoint `GET /loki/api/v1/split_query` is introduced that takes an `splits` parameter and the same parameters as the [/loki/api/v1/query_range](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) endpoint.

The `splits` parameter defines the number of desired splits. The API is allowed to return fewer splits than requested.

The response body is JSON encoded

```json
{ 
  "resultType": "matrix" | "streams" | "vector",
  “links”: [
  {

>     “rel”: “???”,

Author: This is from an old design doc on pagination which was based on GitHub's [pagination API](https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28#using-link-headers).
